### ISSUE TITLE
Precision Differences in Using `dropout` and `linear`

### ISSUE THREAD

**Issue by SORT2024** [pytorch/pytorch#133006]
### 🐛 Describe the bug

**Description:**

I have encountered some precision differences when using the `dropout` and `linear` in the following way.

**Code to Reproduce:**

```python
import torch.nn.functional as f
import torch

# Precision error is calculated based on the Chebyshev distance.

args = torch.load('dropout.pt')

output = f.dropout(args['parameter:0'], args['parameter:1'], args['parameter:2'], args['parameter:3'])

args = torch.load('linear.pt')

output = f.linear(output, args['parameter:1'], args['parameter:2'])
```

**Precision Differences:**

- `dropout`: 
  - CPU: -397.17065
  - GPU: -397.17065
  - Precision Differences: 0.
- `linear`: 
  - CPU: -43764688.
  - GPU: -43764812.
  - Precision Differences: 124.

**System Info:**
- PyTorch Version: 2.3.1+cu121
- OS: Ubuntu 20.04.3 LTS
- Python version: Python 3.8.0
- CUDA version: 12.1

**Expected Behavior:**
The calculation results of every API are consistent across CPU and GPU, with precision differences less than 1e-3.

**Actual Behavior:**
The CPU and GPU results of the second API `linear` show differences greater than the accepted threshold 1e-3.

**Additional Context:**
[2200 pt files](https://drive.google.com/drive/folders/17CIzezLLKD86a4XBTntrzAi2GRg7GU-i?usp=sharing)

Thank you for your attention to this matter. Please let me know if any further information is required.

### Versions

PyTorch version: 2.3.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.0 (default, Nov  6 2019, 21:49:08)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-100-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 10.1.243
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti
GPU 2: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 550.54.14
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      45 bits physical, 48 bits virtual
CPU(s):                             80
On-line CPU(s) list:                40,58,63,66
Off-line CPU(s) list:               0-39,41-57,59-62,64,65,67-79
Thread(s) per core:                 0
Core(s) per socket:                 40
Socket(s):                          2
NUMA node(s):                       2
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              106
Model name:                         Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
Stepping:                           6
CPU MHz:                            2199.999
BogoMIPS:                           4399.99
L1d cache:                          3.8 MiB
L1i cache:                          2.5 MiB
L2 cache:                           100 MiB
L3 cache:                           78 MiB
NUMA node0 CPU(s):                  0-39
NUMA node1 CPU(s):                  40-79
Vulnerability Gather data sampling: Vulnerable: No microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT disabled
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves wbnoinvd arat avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==2.3.1
[pip3] torchaudio==2.3.1
[pip3] torchvision==0.18.1
[pip3] triton==2.3.1
[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] torch                     2.3.1                    pypi_0    pypi
[conda] torchaudio                2.3.1                    pypi_0    pypi
[conda] torchvision               0.18.1                   pypi_0    pypi
[conda] triton                    2.3.1                    pypi_0    pypi

**Comment by albanD** [comment]
> with precision differences less than 1e-3.

I'm not sure why you expect that? fp32 computation has ~6 digits of precision. Meaning the RELATIVE precision is around 1e-6. with absolutely no guarantee in the ABSOLUTE precision (as you show in your example, with arbitrarily large numbers, you can get the absolute error arbitrarily bad).

**Comment by albanD** [comment]
From looking at all these other issues closed above, they all have the same relative error <1e-6 which is what we expect to see.

**Comment by SORT-1** [comment]
Thank you for the feedback.

We've observed relative differences larger than the expected threshold of 1e-6 in the reported issue https://github.com/pytorch/pytorch/issues/132997 and https://github.com/pytorch/pytorch/issues/133001.

(As a reminder, we calculate relative difference using the formula |x - y| / |x|, where x and y represent the results of CPU and GPU computations. Could you please confirm whether this is consistent with your calculation method?)

**Issue 132997:**

- `batch_norm`: 
  - CPU: -4640.65332
  - GPU: -4640.65234
  - Absolute Differences: 0.00098
  - Relative Differences: 2.11177e-7
- `relu`: 
  - CPU: 2574.08203
  - GPU: 2574.08130
  - Absolute Differences: 0.00073
  - Relative Differences: 2.83596e-7
- `conv2d`: 
  - CPU: 266043.65625
  - GPU: 266025.71875
  - Absolute Differences: 17.93750
  - Relative Differences: 6.74231e-5
- `batch_norm`: 
  - CPU: 36164.51172
  - GPU: 36167.98438
  - Absolute Differences: 3.47266
  - Relative Differences: 9.60240e-5

**Issue 133001:**

- `batch_norm`: 
  - CPU: -3557.41479
  - GPU: -3557.41382
  - Absolute Differences: 0.00098
  - Relative Differences: 2.75481e-7
- `relu`: 
  - CPU: 2807.42822
  - GPU: 2807.42749
  - Absolute Differences: 0.00073
  - Relative Differences: 3.49074e-7
- `max_pool2d`: 
  - CPU: 2807.42822
  - GPU: 2807.42749
  - Absolute Differences: 0.00073
  - Relative Differences: 3.49074e-7
- `conv2d`: 
  - CPU: -457744.15625
  - GPU: -457730.50000
  - Absolute Differences: 13.65625
  - Relative Differences: 2.98338e-5

**Comment by SORT-1** [comment]
Actually, in our experiments, we have found quite a few similar cases with relative differences. There are some examples in our testing that have both absolute and relative differences:

**Example 1:**

- `conv2d`: 
  - CPU: -3.72656
  - GPU: -3.99239
  - Absolute Differences: 0.26583
  - Relative Differences: 0.07133
- `__add__`: 
  - CPU: -0.94141
  - GPU: -1.04297
  - Absolute Differences: 0.10156
  - Relative Differences: 0.10788

**Example 2:**

- `__add__`: 
  - CPU: 47684.41797
  - GPU: 47684.41797
  - Absolute Differences: 0.
  - Relative Differences: 0.
- `__add__`: 
  - CPU: 68360.81250
  - GPU: 68360.81250
  - Absolute Differences: 0.
  - Relative Differences: 0.
- `conv2d`: 
  - CPU: -36.25000
  - GPU: -39.58242
  - Absolute Differences: 3.33242
  - Relative Differences: 0.09193
- `batch_norm`: 
  - CPU: 0.71210
  - GPU: 0.95693
  - Absolute Differences: 0.24483
  - Relative Differences: 0.34382


We’re concerned about the `conv2d` and `__add__` operations in Example 1, and the `conv2d` and `batch_norm` operations in Example 2, where the relative differences between CPU and GPU results exceed 1e-6. Would this be considered acceptable within the expected precision?

**Comment by SORT-1** [comment]
Additionally, some absolute differences can become significant when they accumulate over many computations. Based on our collaboration with some companies, this accumulation of differences will draw their attention as the differences will impact the consistency of the results produced by different devices. For instance, according to the data from issue https://github.com/pytorch/pytorch/issues/132997, we can observe how these differences accumulate and lead to the absolute differences become bigger and relative differences between CPU and GPU results exceed 1e-6.

**Comment by albanD** [comment]
Thanks for your answer:

> We've observed relative differences larger than the expected threshold of 1e-6 in the reported issue https://github.com/pytorch/pytorch/issues/132997 and https://github.com/pytorch/pytorch/issues/133001.

The 1e-6 difference is for every single floating point operation done with fp32 precision.
This means that any op performing a reduction can lead to significantly larger relative error depending on the size of the dimension being reduced. In particular, operations like mm and convolutions are expected to lead to larger differences if the input size is large enough.

> `__add__` error

Can you share which exact issue this relates to? `__add__` being element-wise should not exhibit any more than 1e-6 relative error for sure. Checking the issues I see a 0 relative error for add but maybe I missed one?


> Additionally, some absolute differences can become significant when they accumulate over many computations.

This is indeed a major concern for machine learning as a field. The general strategy being used is to have networks that are resilient to such small errors and remain stable in the presence of floating point errors.
I think that non-reproducibility across hardware is a widely accepted thorn with no solution on our end (it stems from the IEEE floating point standard).

**Comment by SORT-1** [comment]
> Can you share which exact issue this relates to?

The Example 1 comes from the case where we called `conv2d` and `__add__` in the following way. You can find the input and output information in this [link](https://drive.google.com/drive/folders/15GU7Y0-pfNTeu4nz-gr7tQdfF9FC4x8X?usp=sharing).

And thank you very much for the elaboration you've provided above.
